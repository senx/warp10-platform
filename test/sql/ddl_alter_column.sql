-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.

CREATE TABLE alter_test(time timestamptz, temp float, color varchar(10));

-- create hypertable with two chunks
SELECT create_hypertable('alter_test', 'time', 'color', 2, chunk_time_interval => 2628000000000);

INSERT INTO alter_test VALUES ('2017-01-20T09:00:01', 17.5, 'blue'),
                              ('2017-01-21T09:00:01', 19.1, 'yellow'),
                              ('2017-04-20T09:00:01', 89.5, 'green'),
                              ('2017-04-21T09:00:01', 17.1, 'black');
SELECT * FROM test.show_columns('alter_test');
SELECT * FROM test.show_columnsp('_timescaledb_internal.%chunk');

-- show the column name and type of the partitioning dimension in the
-- metadata table
SELECT * FROM _timescaledb_catalog.dimension;

EXPLAIN (costs off)
SELECT * FROM alter_test WHERE time > '2017-05-20T10:00:01';

-- rename column and change its type
ALTER TABLE alter_test RENAME COLUMN time TO time_us;
--converting timestamptz->timestamp should happen under UTC
SET timezone = 'UTC';
ALTER TABLE alter_test ALTER COLUMN time_us TYPE timestamp;
RESET timezone;
ALTER TABLE alter_test RENAME COLUMN color TO colorname;
\set ON_ERROR_STOP 0
-- Changing types on hash-partitioned columns is not safe for some
-- types and is therefore blocked.
ALTER TABLE alter_test ALTER COLUMN colorname TYPE text;
\set ON_ERROR_STOP 1

SELECT * FROM test.show_columns('alter_test');
SELECT * FROM test.show_columnsp('_timescaledb_internal.%chunk');

-- show that the metadata has been updated
SELECT * FROM _timescaledb_catalog.dimension;

-- constraint exclusion should still work with updated column
EXPLAIN (costs off)
SELECT * FROM alter_test WHERE time_us > '2017-05-20T10:00:01';

\set ON_ERROR_STOP 0
-- verify that we cannot change the column type to something incompatible
ALTER TABLE alter_test ALTER COLUMN colorname TYPE varchar(3);
-- conversion that messes up partitioning fails
ALTER TABLE alter_test ALTER COLUMN time_us TYPE timestamptz USING time_us::timestamptz+INTERVAL '1 year';
-- dropping column that messes up partiitoning fails
ALTER TABLE alter_test DROP COLUMN colorname;
--ONLY blocked
ALTER TABLE ONLY alter_test RENAME COLUMN colorname TO colorname2;
ALTER TABLE ONLY alter_test ALTER COLUMN colorname TYPE varchar(10);
\set ON_ERROR_STOP 1

CREATE TABLE alter_test_bigint(time bigint, temp float);
SELECT create_hypertable('alter_test_bigint', 'time', chunk_time_interval => 2628000000000);

\set ON_ERROR_STOP 0
-- Changing type of time dimension to a non-supported type
-- shall not be allowed
ALTER TABLE alter_test_bigint
ALTER COLUMN time TYPE TEXT;
-- dropping open time dimension shall not be allowed.
ALTER TABLE alter_test_bigint
DROP COLUMN time;
\set ON_ERROR_STOP 1
