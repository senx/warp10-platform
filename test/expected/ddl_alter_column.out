-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.
CREATE TABLE alter_test(time timestamptz, temp float, color varchar(10));
-- create hypertable with two chunks
SELECT create_hypertable('alter_test', 'time', 'color', 2, chunk_time_interval => 2628000000000);
NOTICE:  adding not-null constraint to column "time"
    create_hypertable    
-------------------------
 (1,public,alter_test,t)
(1 row)

INSERT INTO alter_test VALUES ('2017-01-20T09:00:01', 17.5, 'blue'),
                              ('2017-01-21T09:00:01', 19.1, 'yellow'),
                              ('2017-04-20T09:00:01', 89.5, 'green'),
                              ('2017-04-21T09:00:01', 17.1, 'black');
SELECT * FROM test.show_columns('alter_test');
 Column |           Type           | NotNull 
--------+--------------------------+---------
 time   | timestamp with time zone | t
 temp   | double precision         | f
 color  | character varying        | f
(3 rows)

SELECT * FROM test.show_columnsp('_timescaledb_internal.%chunk');
                Relation                | Kind | Column |       Column type        | NotNull 
----------------------------------------+------+--------+--------------------------+---------
 _timescaledb_internal._hyper_1_1_chunk | r    | time   | timestamp with time zone | t
 _timescaledb_internal._hyper_1_1_chunk | r    | temp   | double precision         | f
 _timescaledb_internal._hyper_1_1_chunk | r    | color  | character varying        | f
 _timescaledb_internal._hyper_1_2_chunk | r    | time   | timestamp with time zone | t
 _timescaledb_internal._hyper_1_2_chunk | r    | temp   | double precision         | f
 _timescaledb_internal._hyper_1_2_chunk | r    | color  | character varying        | f
 _timescaledb_internal._hyper_1_3_chunk | r    | time   | timestamp with time zone | t
 _timescaledb_internal._hyper_1_3_chunk | r    | temp   | double precision         | f
 _timescaledb_internal._hyper_1_3_chunk | r    | color  | character varying        | f
(9 rows)

-- show the column name and type of the partitioning dimension in the
-- metadata table
SELECT * FROM _timescaledb_catalog.dimension;
 id | hypertable_id | column_name |       column_type        | aligned | num_slices | partitioning_func_schema | partitioning_func  | interval_length | integer_now_func_schema | integer_now_func 
----+---------------+-------------+--------------------------+---------+------------+--------------------------+--------------------+-----------------+-------------------------+------------------
  1 |             1 | time        | timestamp with time zone | t       |            |                          |                    |   2628000000000 |                         | 
  2 |             1 | color       | character varying        | f       |          2 | _timescaledb_internal    | get_partition_hash |                 |                         | 
(2 rows)

EXPLAIN (costs off)
SELECT * FROM alter_test WHERE time > '2017-05-20T10:00:01';
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Append
   ->  Index Scan using _hyper_1_2_chunk_alter_test_time_idx on _hyper_1_2_chunk
         Index Cond: ("time" > 'Sat May 20 10:00:01 2017 PDT'::timestamp with time zone)
   ->  Index Scan using _hyper_1_3_chunk_alter_test_time_idx on _hyper_1_3_chunk
         Index Cond: ("time" > 'Sat May 20 10:00:01 2017 PDT'::timestamp with time zone)
(5 rows)

-- rename column and change its type
ALTER TABLE alter_test RENAME COLUMN time TO time_us;
--converting timestamptz->timestamp should happen under UTC
SET timezone = 'UTC';
ALTER TABLE alter_test ALTER COLUMN time_us TYPE timestamp;
RESET timezone;
ALTER TABLE alter_test RENAME COLUMN color TO colorname;
\set ON_ERROR_STOP 0
-- Changing types on hash-partitioned columns is not safe for some
-- types and is therefore blocked.
ALTER TABLE alter_test ALTER COLUMN colorname TYPE text;
ERROR:  cannot change the type of a hash-partitioned column
\set ON_ERROR_STOP 1
SELECT * FROM test.show_columns('alter_test');
  Column   |            Type             | NotNull 
-----------+-----------------------------+---------
 time_us   | timestamp without time zone | t
 temp      | double precision            | f
 colorname | character varying           | f
(3 rows)

SELECT * FROM test.show_columnsp('_timescaledb_internal.%chunk');
                Relation                | Kind |  Column   |         Column type         | NotNull 
----------------------------------------+------+-----------+-----------------------------+---------
 _timescaledb_internal._hyper_1_1_chunk | r    | time_us   | timestamp without time zone | t
 _timescaledb_internal._hyper_1_1_chunk | r    | temp      | double precision            | f
 _timescaledb_internal._hyper_1_1_chunk | r    | colorname | character varying           | f
 _timescaledb_internal._hyper_1_2_chunk | r    | time_us   | timestamp without time zone | t
 _timescaledb_internal._hyper_1_2_chunk | r    | temp      | double precision            | f
 _timescaledb_internal._hyper_1_2_chunk | r    | colorname | character varying           | f
 _timescaledb_internal._hyper_1_3_chunk | r    | time_us   | timestamp without time zone | t
 _timescaledb_internal._hyper_1_3_chunk | r    | temp      | double precision            | f
 _timescaledb_internal._hyper_1_3_chunk | r    | colorname | character varying           | f
(9 rows)

-- show that the metadata has been updated
SELECT * FROM _timescaledb_catalog.dimension;
 id | hypertable_id | column_name |         column_type         | aligned | num_slices | partitioning_func_schema | partitioning_func  | interval_length | integer_now_func_schema | integer_now_func 
----+---------------+-------------+-----------------------------+---------+------------+--------------------------+--------------------+-----------------+-------------------------+------------------
  1 |             1 | time_us     | timestamp without time zone | t       |            |                          |                    |   2628000000000 |                         | 
  2 |             1 | colorname   | character varying           | f       |          2 | _timescaledb_internal    | get_partition_hash |                 |                         | 
(2 rows)

-- constraint exclusion should still work with updated column
EXPLAIN (costs off)
SELECT * FROM alter_test WHERE time_us > '2017-05-20T10:00:01';
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Append
   ->  Seq Scan on _hyper_1_2_chunk
         Filter: (time_us > 'Sat May 20 10:00:01 2017'::timestamp without time zone)
   ->  Seq Scan on _hyper_1_3_chunk
         Filter: (time_us > 'Sat May 20 10:00:01 2017'::timestamp without time zone)
(5 rows)

\set ON_ERROR_STOP 0
-- verify that we cannot change the column type to something incompatible
ALTER TABLE alter_test ALTER COLUMN colorname TYPE varchar(3);
ERROR:  cannot change the type of a hash-partitioned column
-- conversion that messes up partitioning fails
ALTER TABLE alter_test ALTER COLUMN time_us TYPE timestamptz USING time_us::timestamptz+INTERVAL '1 year';
ERROR:  check constraint "constraint_1" is violated by some row
-- dropping column that messes up partiitoning fails
ALTER TABLE alter_test DROP COLUMN colorname;
ERROR:  cannot drop column named in partition key
--ONLY blocked
ALTER TABLE ONLY alter_test RENAME COLUMN colorname TO colorname2;
ERROR:  inherited column "colorname" must be renamed in child tables too
ALTER TABLE ONLY alter_test ALTER COLUMN colorname TYPE varchar(10);
ERROR:  ONLY option not supported on hypertable operations
\set ON_ERROR_STOP 1
CREATE TABLE alter_test_bigint(time bigint, temp float);
SELECT create_hypertable('alter_test_bigint', 'time', chunk_time_interval => 2628000000000);
NOTICE:  adding not-null constraint to column "time"
       create_hypertable        
--------------------------------
 (2,public,alter_test_bigint,t)
(1 row)

\set ON_ERROR_STOP 0
-- Changing type of time dimension to a non-supported type
-- shall not be allowed
ALTER TABLE alter_test_bigint
ALTER COLUMN time TYPE TEXT;
ERROR:  cannot change data type of hypertable column "time" from bigint to text
-- dropping open time dimension shall not be allowed.
ALTER TABLE alter_test_bigint
DROP COLUMN time;
ERROR:  cannot drop column named in partition key
\set ON_ERROR_STOP 1
